src/Physics/NuclearState/NucleusGenINCL.cxx

```diff
@@ -0,0 +1,336 @@
+
+  // handle the addition of the recoil nucleon
+  // TODO:  INCL has it own SRC model
+//  if ( fSecondEmitter ) fSecondEmitter -> ProcessEventRecord( evrec ) ;
+
+  // add a recoiled nucleus remnant
+  this->setTargetNucleusRemnant(evrec);
+  LOG("NucleusGenINCL", pINFO) << "Adding final state nucleus";
+}
+
+//___________________________________________________________________________
+//  using INCL model to get the position and momentum of 
+//  Hit  nucleon
+void NucleusGenINCL::setInitialStateVertex(GHepRecord * evrec) const{
+
+  INCLNucleus *incl_nucleus = INCLNucleus::Instance();
+
+// generate a vtx and set it to all GHEP physical particles
+  Interaction * interaction = evrec->Summary();
+  GHepParticle * nucltgt = evrec->TargetNucleus();
+  TVector3 vtx(9999999.,999999.,999999.);
+  if(!nucltgt){
+    vtx.SetXYZ(0.,0.,0.);
+  }else{
+    double A = nucltgt->A();
+ //   vtx = GenerateVertex(interaction,A);
+ 	
+  const ProcessInfo & proc_info = interaction->ProcInfo();
+  bool is_coh = proc_info.IsCoherentProduction() || proc_info.IsCoherentElastic();
+  bool is_ve  = proc_info.IsInverseMuDecay() ||
+    proc_info.IsIMDAnnihilation() ||
+    proc_info.IsNuElectronElastic() ||
+    proc_info.IsGlashowResonance() ||
+    proc_info.IsPhotonResonance() ||
+    proc_info.IsPhotonCoherent();
+
+
+  if(is_coh||is_ve) {
+    // ** For COH or ve- set a vertex positon on the nuclear boundary
+    //
+  //  LOG("Vtx", pINFO)  << "Setting vertex on the nuclear boundary";
+  //  double phi      = 2*kPi * rnd->RndFsi().Rndm();
+  //  double cosphi   = TMath::Cos(phi);
+  //  double sinphi   = TMath::Sin(phi);
+  //  double costheta = -1 + 2 * rnd->RndFsi().Rndm();
+  //  double sintheta = TMath::Sqrt(1-costheta*costheta);
+  //  vtx.SetX(R*sintheta*cosphi);
+  //  vtx.SetY(R*sintheta*sinphi);
+  //  vtx.SetZ(R*costheta);
+  }
+  else {
+    vtx = incl_nucleus->getHitNucleonPosition();
+  }
+
+  }
+
+  LOG("NucleusGenINCL", pINFO) << "Position";
+  vtx.Print();
+  // Copy the vertex info to the particles already in the event  record
+  //
+  TObjArrayIter piter(evrec);
+  GHepParticle * p = 0;
+  while( (p = (GHepParticle *) piter.Next()) )
+  {
+    if(pdg::IsPseudoParticle(p->Pdg())) continue;
+    if(pdg::IsIon           (p->Pdg())) continue;
+
+    LOG("NucleusGenINCL", pINFO) << "Setting vertex position for: " << p->Name();
+    p->SetPosition(vtx.x(), vtx.y(), vtx.z(), 0.);
+  }
+
+
+
+}
+
+void NucleusGenINCL::setInitialStateMomentum(GHepRecord * evrec) const{
+  Interaction *  interaction = evrec       -> Summary();
+  InitialState * init_state  = interaction -> InitStatePtr();
+  Target *       tgt         = init_state  -> TgtPtr();
+
+  // do nothing for non-nuclear targets
+  if(!tgt->IsNucleus()) return;
+
+  TLorentzVector * p4 = tgt->HitNucP4Ptr();
+
+  // do nothing if the struct nucleon 4-momentum was set (eg as part of the
+  // initial state selection)
+  if(p4->Px()>0 || p4->Py()>0 || p4->Pz()>0) return;
```

Comment:
what if all three are non-zero but negative?  Perhaps `p4->Vect()->Mag2() > 0` trading three multiplies and one comparison, against 3 `TMath::Abs()` and three comparisons?